src/data/reports/glass-data-submission/GLASSDataSubmissionDefaultRepository.ts

```diff
@@ -825,23 +898,143 @@ export class GLASSDataSubmissionDefaultRepository implements GLASSDataSubmission
         });
     }
 
-    private async duplicateProgram(program: ApprovalIds, items: GLASSDataSubmissionItemIdentifier[]) {
-        await promiseMap(items, async item => {
-            const programEvents = await this.getProgramEvents(program.id, item.orgUnit ?? "");
+    private async duplicateTrackerProgram(
+        program: ApprovalIds,
+        items: GLASSDataSubmissionItemIdentifier[]
+    ): Promise<void> {
+        const approvedProgramStages = await this.getProgramStages(program.approvedId);
+        const programStages = await this.getProgramStages(program.id);
+
+        _.forEach(items, async item => {
+            const trackedEntities = await this.getTrackedEntities(program.id, item.orgUnit);
+
+            if (!_.isEmpty(trackedEntities)) {
+                const approvedTrackedEntities = await this.getTrackedEntities(program.approvedId, item.orgUnit);
+
+                !_.isEmpty(approvedTrackedEntities) &&
+                    this.api.post(
+                        "/tracker",
+                        { importStrategy: "DELETE" },
+                        { trackedEntities: approvedTrackedEntities }
+                    );
+
+                const trackedEntitiesToPost: TrackedEntity[] = trackedEntities.map(trackedEntity => ({
+                    ...trackedEntity,
+                    trackedEntity: "",
+                    programOwners: trackedEntity.programOwners.map(programOwner => ({
+                        ...programOwner,
+                        program: program.approvedId,
+                    })),
+                    enrollments: trackedEntity.enrollments.map(enrollment => ({
+                        ...enrollment,
+                        enrollment: "",
+                        trackedEntity: "",
+                        program: program.approvedId,
+                        events: enrollment.events.map(event => {
+                            const programStageName = programStages.find(
+                                programStage => programStage.id === event.programStage
+                            )?.name;
+                            const programStageApproved =
+                                approvedProgramStages.find(
+                                    approvedProgramStage => approvedProgramStage.name === programStageName
+                                )?.id ?? "";
+
+                            return {
+                                ...event,
+                                event: "",
+                                trackedEntity: "",
+                                program: program.approvedId,
+                                programStage: programStageApproved,
+                            };
+                        }),
+                    })),
+                }));
+
+                return await this.api.post("/tracker", {}, { trackedEntities: trackedEntitiesToPost }).getData();
```

Comment:
do we need to mention {async : true, importStrategy : "CREATE_AND_UPDATE"} here? Are they the default values? Better to post asynchronously as the number of trackedEntities could be very larhe.

Reply:
got it!